dst: Update `GetProfile`'s stream when pod associated to HostPort lookup changes

alpeb opened the pull request:
Followup to #11328

Implements a new pod watcher, instantiated along the other ones in the Destination server. It also watches on Servers and carries all the logic from ServerWatcher, which has now been decommissioned.

The `CreateAddress()` function has been moved into a function of the PodWatcher, because now we're calling it on every update given the pod associated to an ip:port might change and we need to regenerate the Address object. That function also takes care of capturing opaque protocol info from associated Servers, which is not new and had some logic that was duped in the now defunct ServerWatcher. `getAnnotatedOpaquePorts()` got also moved for similar reasons.

Other things to note about PodWatcher:

- It publishes a new pair of metrics `ip_port_subscribers` and `ip_port_updates` leveraging the framework in `prometheus.go`.
- The complexity in `updatePod()` is due to only send stream updates when there are changes in the pod's readiness, to avoid sending duped messages on every pod lifecycle event.
- 
Finally, endpointProfileTranslator's `endpoint` (*pb.WeightedAddr) not being a static object anymore, the `Update()` function now receives an Address that allows it to rebuild the endpoint on the fly (and so `createEndpoint()` was converted into a method of endpointProfileTranslator).

------------------------

## Original PR description

(What follows is the original description of this PR's first approach, which evolved as described above after taking into account all the feedback).

Implements a new pod watcher, instantiated along the other ones in the Destination server. It's generic enough to catch all pod events in the cluster, so it's up to the subscribers to filter out the ones they're interested in, and to set up any metrics.

The `endointProfileTranslator` was refactored so it subscribes to the pod watcher events. Handling of Server subscriptions are now also handled by it, which are recycled whenever the pod changes.

A new gauge metric `host_port_subscribers` has been created, tracking the number of subscribers for a given HostIP+port combination.

### Other Changes

- Moved the `server.createAddress` method into a static function in `endpoints_watcher.go`, for better reusability.
- The "Return profile for host port pods" test introduced in #11328 was extended to track the ensuing events after a pod is deleted and then recreated (:taco: to @adleong for the test).
- Given that test consumes multiple events, we had to change the `profileStream` test helper to have `GetProfile` run in a goroutine so that it keeps on running. This implies the stream now needs to be manually canceled and access to the stream updates have to be guarded with a mutex.

mateiidavid commented on controller/api/destination/hostport_adaptor.go:
```diff
@@ -0,0 +1,257 @@
+}
+
+func (pt *hostPortAdaptor) updateAddress(pod *corev1.Pod) {
+	go func() {
```
Should we leave concurrency to the caller here? I feel like a go routine here might make it more difficult to troubleshoot.

alpeb replied:
It seemed more succinct to me to do it in the common method. Can you clarify how the alternative would help?

alpeb replied:
Ok I ended up moving the goroutines creation to the consumers in my latest refactoring, like you suggested. Those functions explicitly ask for non-blocking code in their godocs, so it's more explicit to handle that right in there.

alpeb replied:
Yep that makes sense, thanks for the clarification!

mateiidavid reviewed (commented):
Looks great! Think you broke the problem down really well too, it was easy for me to jump in with no context. I left a few comments, I need to understand the adaptor a bit better. Using the pattern here is difficult because it would otherwise be hard to slot the pod watcher into the endpoint translator? (e.g. running a watcher that can propagate updates to endpoint translator listeners). The idea being that the adaptor is the listener that understands pod events?

As far as testing goes, I gave this a try and it works well!

```
INFO[0000] retry_budget:{retry_ratio:0.2 min_retries_per_second:10 ttl:{seconds:10}} endpoint:{addr:{ip:{ipv4:2886860802} port:9999} weight:10000 metric_labels:{key:"control_plane_ns" value:"linkerd"} metric_labels:{key:"deployment" value:"nginx-deployment"} metric_labels:{key:"namespace" value:"default"} metric_labels:{key:"pod" value:"nginx-deployment-674b6bd49b-5tj4r"} metric_labels:{key:"pod_template_hash" value:"674b6bd49b"} metric_labels:{key:"serviceaccount" value:"default"} tls_identity:{dns_like_identity:{name:"default.default.serviceaccount.identity.linkerd.cluster.local"}} protocol_hint:{h2:{}}}
//
// RESTART POD
//
INFO[0000] 
INFO[0008] retry_budget:{retry_ratio:0.2 min_retries_per_second:10 ttl:{seconds:10}} endpoint:{addr:{ip:{ipv4:2886860802} port:9999} weight:10000}
INFO[0008]
INFO[0012] retry_budget:{retry_ratio:0.2 min_retries_per_second:10 ttl:{seconds:10}} endpoint:{addr:{ip:{ipv4:2886860802} port:9999} weight:10000 metric_labels:{key:"control_plane_ns" value:"linkerd"} metric_labels:{key:"deployment" value:"nginx-deployment"} metric_labels:{key:"namespace" value:"default"} metric_labels:{key:"pod" value:"nginx-deployment-674b6bd49b-6ngm2"} metric_labels:{key:"pod_template_hash" value:"674b6bd49b"} metric_labels:{key:"serviceaccount" value:"default"} tls_identity:{dns_like_identity:{name:"default.default.serviceaccount.identity.linkerd.cluster.local"}} protocol_hint:{h2:{}}}
//
// RESTART POD
//
INFO[0036] retry_budget:{retry_ratio:0.2 min_retries_per_second:10 ttl:{seconds:10}} endpoint:{addr:{ip:{ipv4:2886860802} port:9999} weight:10000}
INFO[0036]
INFO[0041] retry_budget:{retry_ratio:0.2 min_retries_per_second:10 ttl:{seconds:10}} endpoint:{addr:{ip:{ipv4:2886860802} port:9999} weight:10000 metric_labels:{key:"control_plane_ns" value:"linkerd"} metric_labels:{key:"deployment" value:"nginx-deployment"} metric_labels:{key:"namespace" value:"default"} metric_labels:{key:"pod" value:"nginx-deployment-674b6bd49b-gnd6w"} metric_labels:{key:"pod_template_hash" value:"674b6bd49b"} metric_labels:{key:"serviceaccount" value:"default"} tls_identity:{dns_like_identity:{name:"default.default.serviceaccount.identity.linkerd.cluster.local"}} protocol_hint:{h2:{}}}
INFO[0041]
```

The only _issue_ is that we first send the endpoint and _then_ the endpoint information in a subsequent update (i.e. pod metrics). idk if this is a big problem, it's not fresh in my brain how the proxy would handle the connections here, since the IP is the same I assume it just re-uses the connection, it won't tear it down.

alpeb commented on controller/api/destination/watcher/pod_watcher.go:
```diff
@@ -0,0 +1,121 @@
+	// it, and remain responsible to what events to take into account.
+	PodWatcher struct {
+		k8sAPI    *k8s.API
+		handle    cache.ResourceEventHandlerRegistration
```
Actually this made me realize we're not required to tear down these handlers, so I'm just gonna get rid of this field altogether to simplify.

alpeb commented on controller/api/destination/watcher/pod_watcher.go:
```diff
@@ -0,0 +1,121 @@
+	for i, e := range pw.listeners {
+		if e == listener {
+			n := len(pw.listeners)
+			pw.listeners[i] = pw.listeners[n-1]
+			pw.listeners[n-1] = nil
+			pw.listeners = pw.listeners[:n-1]
```
Copy-pasted from the server watcher I believe :-P

alpeb commented on controller/api/destination/watcher/pod_watcher.go:
```diff
@@ -0,0 +1,121 @@
+	pw.mu.RLock()
+	defer pw.mu.RUnlock()
+
+	pod := obj.(*corev1.Pod)
```
I missed that , good point :+1:

alpeb commented on controller/api/destination/watcher/pod_watcher.go:
```diff
@@ -0,0 +1,121 @@
+
+	pod := obj.(*corev1.Pod)
+	pw.log.Tracef("Addded pod %s.%s", pod.Name, pod.Namespace)
+	for _, l := range pw.listeners {
+		l.Update(pod)
+	}
```
We could call `DeepCopy` in each of these handlers, but I'm not sure if it's worth it. And it seems we haven't done that in the other watchers in the destination controller?

alpeb commented:
Thanks for the detailed review @mateiidavid.

> Using the pattern here is difficult because it would otherwise be hard to slot the pod watcher into the endpoint translator? (e.g. running a watcher that can propagate updates to endpoint translator listeners). The idea being that the adaptor is the listener that understands pod events?

You're right inquire about the logic behind that separation. I think it made sense at some point during my implementation process, but I see no reason for that anymore. So I'll be coalescing all the logic into a refactored `endpointProfileTranslator`.

> The only issue is that we first send the endpoint and then the endpoint information in a subsequent update (i.e. pod metrics). idk if this is a big problem, it's not fresh in my brain how the proxy would handle the connections here, since the IP is the same I assume it just re-uses the connection, it won't tear it down.

I think the reason behind those separate messages is that after deleting the pod the hostip:port no longer has a backing pod so the following message only sends the IP. And then when the new pod becomes ready a new message is sent with all the pod's info (metric labels, identity and protocol hint).
This seemed to behave well from my cursory testing, but I'd love to get more confirmation that this is a right behavior.

mateiidavid commented on controller/api/destination/hostport_adaptor.go:
```diff
@@ -0,0 +1,257 @@
+		return fmt.Errorf("failed to create endpoint: %w", err)
+	}
+	pt.endpoint = endpoint
+	pt.log.Debugf("Sync for endpoint %s", pt.endpoint)
```
nit: will the endpoint have an underlying string representation? Is it better to use [`pt.endpoint.Addr.String()`](https://github.com/linkerd/linkerd2-proxy-api/blob/main/go/destination/destination.pb.go#L305C2-L305C7)?

olix0r commented:
This seems big enough that I think we ought to hold merging this until we've branched release/stable-2.14.

adleong commented:
There are few things here that deviate from the patterns established in other watchers/translators:
* subscribing to watches in a translator
* having a watcher broadcast ALL events to all listeners without filtering

These are surprising and make this code difficult for me to follow.  In particular, the logic in the endpointProfileTransformer has become very complex.  I had something different in mind:

Rather than having the PodWatcher broadcast all pod events to all listeners, instead have it contain a map of "<ip>:<port>" to PodPublisher, following a similar pattern to the other watches.  PodWatcher would need to register event handlers for both Pods and Servers.  When it gets a pod event, it should update all the registered PodPublishers which match the pod IP or which match the hostport if the pod has one.  When it gets a server event, it should update all the registered PodPublishers which have a pod which matches the server selector.  This moves most of the complexity into the watcher (where it belongs, imo).

This makes PodWatcher follow the same pattern as EndpointsWatcher, but for Pods instead of for Services.

The result of this is that the endpointProfileTranslator just needs to be updated to implement PodListener instead of ServerListener.  All endpointProfile queries (i.e. calls through `subscribeToEndpointProfile`) can use PodWatcher regardless of if the query is for pod network ip or a host network ip.  ServerWatcher becomes unused and can be deleted.

Let me know if this makes sense.  Happy to discuss further.

alpeb commented:
Thanks @adleong for the comprehensive guidelines, it all makes sense to me.
> instead have it contain a map of ":" to PodPublisher

You mean a map of ip:port?

adleong commented:
Yes, I got tripped up by markdown formatting.  Fixed.

adleong commented on controller/api/destination/watcher/pod_watcher.go:
```diff
@@ -0,0 +1,411 @@
+	for _, container := range pod.Spec.Containers {
+		for _, containerPort := range container.Ports {
+			if containerPort.ContainerPort != 0 {
+				if pp, ok := pw.getPodPublisher(pod.Status.PodIP, Port(containerPort.ContainerPort)); ok {
```
Should we be using `getOrNewPodPublisher` here instead so that we're creating a podPublisher with no listeners?  We are, effectively, building our own index.

alpeb replied:
Not sure I follow. Of what use are podPublishers with no listeners? Sounds like a waste of resources to keep podPublishers for all pods in the cluster. If some client becomes interested in the given ip:port, then the `Subscribe()` method will make sure the create the podPublisher (using the pod passed to `Subscribe()` or retrieving it itself as per your previous comment).

adleong replied:
Looking at other watchers like the endpointsWatcher or the profileWatcher, it looks like we populate the publisher map when either a listener subscribes OR when we get an update for that object.  But you're right, we can build the full state when a subscription happens.

alpeb commented on controller/api/destination/watcher/pod_watcher.go:
```diff
@@ -0,0 +1,411 @@
+		go func(pp *podPublisher) {
+			updated := false
+			for _, listener := range pp.listeners {
+				addr, err := CreateAddress(pw.k8sAPI, pw.metadataAPI, pw.defaultOpaquePorts, pp.pod, pp.ip, pp.port)
```
> Is the reason we do this because it's not possible to determine the new opaque ports for a pod just by looking at the updated server because it may have opaque ports which overlap with other servers which match that pod? That sounds right, but it's a subtle point that probably warrants a comment.

Right, the handler carries information limited to the changed Server, and so it only acts to signal that the final state for the relevant pods needs to be recalculated. I'll add a comment to that effect.

> Actually, taking this thinking a step further, we probably need to potentially update ALL podPublishers, even ones which don't match the server's selector because the server's selector may have changed so that it no longer selects certain pods. Those pods which are no longer selected may therefore need their opaque ports updated.
> 
> Taking this all together, I think that any time any server is updated, we need to recalculate all the opaque ports for all the pod publishers. Assuming we're using client-go caches, this shouldn't be too expensive since it won't result in API calls.

Good point. The Server watcher and the lister used in `CreateAddress()` rely on the same informer so indeed, this should result in extra network calls. I'm a little worried though about CPU usage. I think we can be a little smarter in the Server watcher handlers. Currently all events are handled via this single `updateServer()` method. I think it's fine to keep the same logic for the _add_  and _delete_ events, in that only the pods matching the selectors are concerned. For the _update_ event, we can check whether the selector changed, and consider the union of pods that match both the old and new selectors.

adleong replied:
Can we estimate whether this is going to cause problematic CPU usage?  Server updates probably wouldn't happen more than once a second.  On each Server update we would have to evaluate a selector against every pod in the cluster.  Assuming a 10000 pod cluster, doing 10000 selector evaluations every time a server updates doesn't seem like it should be too expensive.  Are there other computational costs we should consider?

alpeb replied:
Ok I wasn't taking into account the frequency of Server updates. I'll leave the logic as is for now and see if I can come up with a way to actually test resource consumption when an update is triggered. Another thing we'd like to avoid is sending dupe messages to every stream on every Server update. I'll add something into the endpointProfileTranslator to avoid that.

alpeb commented:
I Moved the Address/pod object creation logic into the podWatcher, on the `getOrNewPodPublisher()` method, called from `Subscribe()`. The latter receives `service`, `hostname` and `ip`, some of which can be empty depending on the case.

With this change also the auxiliary functions `getIndexedPod()` and `podReceivingTraffic()` were moved into the `watcher` package.

Also the Server updates handler now triggers updates for all podPublishers regardless of the Server selectors. The endpointProfileTranslator is now tracking the last message sent, to avoid triggering unneeded dupe messages.

Other changes:
- Added fields `defaultOpaquePorts`, `k8sAPI` and `metadataAPI` into the podPublisher struct.
- Removed the fields `ip` and `port` from the endpointProfileTranslator, which are not used.

adleong reviewed (approved):
Very nice!

adleong commented:
after thinking about this a bit, yes, I think you're right.  Because the container is available at both the container port (on the pod network) AND the host port (on the host network).

mateiidavid commented on controller/api/destination/watcher/pod_watcher.go:
```diff
@@ -0,0 +1,543 @@
+// Subscribe stops notifying the listener on chages on any pod backing the
+// passed ip:port or its associated protocol config
+func (pw *PodWatcher) Unsubscribe(ip string, port Port, listener PodUpdateListener) {
+	pw.mu.Lock()
+	defer pw.mu.Unlock()
+
+	pw.log.Debugf("Stopping watch on %s:%d", ip, port)
+	pp, ok := pw.getPodPublisher(ip, port)
+	if !ok {
+		pw.log.Errorf("Cannot unsubscribe from unknown ip:port [%s:%d]", ip, port)
+		return
+	}
+	pp.unsubscribe(listener)
```
Any risk of deadlocking here?

```
Unsubscribe():
    pw.mu.Lock()
    defer => lock released when stack is unwinded
    unsubscribe():
        pw.mu.Lock() => lock already held
        defer pw.mu.Unlock()
```

However, we try to acquire the lock in `pp.unsubscribe` . They're both write locks. If we are running in the same thread won't this deadlock?

I put together an example that I think is close to this in concept: https://go.dev/play/p/N30YUDzU7Cj

We could have a read lock here and release it after `getPodPublisher`?

AFAIK Go mutexes aren't recursive/re-entrant (I only found this https://groups.google.com/g/golang-nuts/c/vN5ncBdtkcA).

...but maybe I'm also grossly misinterpreting this code 😅

alpeb replied:
Note that those are different locks: one on `PodWatcher`, the other on `podPublisher` :wink:

mateiidavid commented on controller/api/destination/watcher/pod_watcher.go:
```diff
@@ -0,0 +1,543 @@
+			continue
+		}
+
+		go func(pp *podPublisher) {
+			updated := false
```
should we lock/unlock the mutex while in this go routine? Is it possible for a listener to get concurrently added/removed while we're going through the update logic in this routine?

Tried to put together an example here: https://go.dev/play/p/ADQ-EH3wVER

alpeb replied:
Good catch, we're indeed missing a lock here :+1:

mateiidavid commented on controller/api/destination/watcher/pod_watcher.go:
```diff
@@ -0,0 +1,543 @@
+// since it may be arbitrary we need to look at the corresponding service's
+// Endpoints object to see whether the hostname matches a pod.
+func (pw *PodWatcher) getEndpointByHostname(hostname string, svcID *ServiceID) (*corev1.Pod, error) {
+	ep, err := pw.k8sAPI.Endpoint().Lister().Endpoints(svcID.Namespace).Get(svcID.Name)
```
Should we opt to use the EndpointSlice informer here? Can also be done in a follow-up though, would be helpful to decorate the fn with a todo. My feeling is that we should gradually stop supporting both and instead rely on the newer interface but it's also trivial for this PR 🤷🏻

alpeb replied:
As we rely on `Endpoints` elsewhere as well, I think we should do this change wholesale instead, stopping supporting clusters where `EndpointSlices` is not enabled. We need to check as of what k8s version this is feasible, so I'd rather tackle that separately.

mateiidavid reviewed (commented):
wow! great change, lots of moving parts but it all makes sense and the design is imo really clean.

This is ready to ship from my side. Left two concurrency related questions though.

alpeb commented on controller/api/destination/server.go:
```diff
@@ -467,88 +442,37 @@ func (s *server) subscribeToServiceWithoutContext(
-	var endpoint *pb.WeightedAddr
-	endpoint, err := s.createEndpoint(*address, opaquePorts)
+	var err error
+	ip, err = s.pods.Subscribe(service, hostname, ip, port, translator)
```
I pushed that change, but the linter is rejecting it because `ip` is already defined (as a function parameter). I'm gonna revert the change, even though my local linter version isn't complaning :man_shrugging:

mateiidavid replied:
Ah :( yeah I thought it might've been like this because of some linter complaints. fwiw I have also tried it locally before suggesting the change and it seemed to work. Sounds good to me!

alpeb commented on controller/api/destination/watcher/endpoints_watcher.go:
```diff
@@ -1321,7 +1321,8 @@
 				for _, c := range address.Pod.Spec.Containers {
 					for _, p := range c.Ports {
-						if p.ContainerPort == int32(port) && p.Name == server.Spec.Port.StrVal {
+						if (p.ContainerPort == int32(port) || p.HostPort == int32(port)) &&
```
This alert is bogus because right after the `strconv.Atoi` call mentioned there we're already checking for boundaries.

mateiidavid reviewed (approved):
You have my 🦭 of ✅ !

alpeb commented:
Thanks folks for the great feedback :smile: 
I'm merging this and will shortly followup with an extension of the unit test to cover the bug I fixed (described in the https://github.com/linkerd/linkerd2/pull/11334#issuecomment-1736093592 above).
